src/AddIns/Uno.WinUI.Graphics/GLCanvasElement.shared.cs

```diff
@@ -48,90 +79,174 @@ public abstract partial class GLCanvasElement : UserControl
-	/// should override this method if they need something different.
-	/// </summary>
-	/// <remarks>An exception will be thrown if availableSize is infinite (e.g. if inside a StackPanel).</remarks>
-	protected override partial Size MeasureOverride(Size availableSize);
+	private unsafe void OnLoaded(object sender, RoutedEventArgs routedEventArgs)
+	{
+		_nativeOpenGlWrapper.CreateContext(this);
+		_gl = (GL)_nativeOpenGlWrapper.CreateGLSilkNETHandle();
 
-	/// <summary>
-	/// By default, <see cref="GLCanvasElement"/> uses all the <see cref="finalSize"/> given. Subclasses of <see cref="GLCanvasElement"/>
-	/// should override this method if they need something different.
-	/// </summary>
-	/// <remarks>An exception will be thrown if <see cref="finalSize"/> is infinite (e.g. if inside a StackPanel).</remarks>
-	protected override partial Size ArrangeOverride(Size finalSize);
+		_pixels = Marshal.AllocHGlobal((int)(_width * _height * BytesPerPixel));
 
-	private readonly struct GLStateDisposable : IDisposable
+		using (new GLStateDisposable(this))
+		{
+			_framebuffer = _gl.GenBuffer();
+			_gl.BindFramebuffer(GLEnum.Framebuffer, _framebuffer);
+			{
+				_textureColorBuffer = _gl.GenTexture();
+				_gl.BindTexture(GLEnum.Texture2D, _textureColorBuffer);
+				{
+					_gl.TexImage2D(GLEnum.Texture2D, 0, InternalFormat.Rgb, _width, _height, 0, GLEnum.Rgb, GLEnum.UnsignedByte, (void*)0);
+					_gl.TexParameterI(GLEnum.Texture2D, GLEnum.TextureMinFilter, (uint)GLEnum.Linear);
+					_gl.TexParameterI(GLEnum.Texture2D, GLEnum.TextureMagFilter, (uint)GLEnum.Linear);
+					_gl.FramebufferTexture2D(GLEnum.Framebuffer, FramebufferAttachment.ColorAttachment0, GLEnum.Texture2D, _textureColorBuffer, 0);
+				}
+				_gl.BindTexture(GLEnum.Texture2D, 0);
+
+				_renderBuffer = _gl.GenRenderbuffer();
+				_gl.BindRenderbuffer(GLEnum.Renderbuffer, _renderBuffer);
+				{
+					_gl.RenderbufferStorage(GLEnum.Renderbuffer, InternalFormat.Depth24Stencil8, _width, _height);
+					_gl.FramebufferRenderbuffer(GLEnum.Framebuffer, GLEnum.DepthStencilAttachment, GLEnum.Renderbuffer, _renderBuffer);
+				}
+				_gl.BindRenderbuffer(GLEnum.Renderbuffer, 0);
+
+				if (_gl.CheckFramebufferStatus(GLEnum.Framebuffer) != GLEnum.FramebufferComplete)
+				{
+					throw new InvalidOperationException("Offscreen framebuffer is not complete");
+				}
+
+				Init(_gl);
+			}
+			_gl.BindFramebuffer(GLEnum.Framebuffer, 0);
+		}
+
+		Invalidate();
+	}
+
+	private void OnUnloaded(object sender, RoutedEventArgs routedEventArgs)
 	{
-		private readonly GL _gl;
-		private readonly int _oldArrayBuffer;
-		private readonly int _oldVertexArray;
-		private readonly int _oldFramebuffer;
-		private readonly int _oldTextureColorBuffer;
-		private readonly int _oldRbo;
-		private readonly bool _depthTestEnabled;
-		private readonly bool _depthTestMask;
-		private readonly int[] _oldViewport = new int[4];
+		Debug.Assert(_gl is not null); // because OnLoaded creates _gl
```

Comment:
At least on WinAppSDK, this assert is wrong. There are cases where Unloaded is fired before Loaded.

Reply:
Let's let it be for now until this becomes a problem. My understanding is that problems only happen in extreme edge cases and the common cases should be fine.